Python/fileutils.c

```diff
@@ -1097,13 +1115,72 @@ _Py_attribute_data_to_stat(BY_HANDLE_FILE_INFORMATION *info, ULONG reparse_tag,
     result->st_reparse_tag = reparse_tag;
     if (info->dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
         reparse_tag == IO_REPARSE_TAG_SYMLINK) {
-        /* first clear the S_IFMT bits */
-        result->st_mode ^= (result->st_mode & S_IFMT);
-        /* now set the bits that make this a symlink */
-        result->st_mode |= S_IFLNK;
+        /* set the bits that make this a symlink */
+        result->st_mode = (result->st_mode & ~S_IFMT) | S_IFLNK;
     }
     result->st_file_attributes = info->dwFileAttributes;
 }
+
+void
+_Py_stat_basic_info_to_stat(FILE_STAT_BASIC_INFORMATION *info,
+                            struct _Py_stat_struct *result)
+{
+    /* set values here to match WIN32_STATX_BASIC_STATS defined in posixmodule.c */
+    memset(result, 0, sizeof(*result));
+    result->st_mode = attributes_to_mode(info->FileAttributes);
+    result->st_size = info->EndOfFile.QuadPart;
+    LARGE_INTEGER_to_time_t_nsec(&info->CreationTime, &result->st_btime, &result->st_btime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&info->ChangeTime, &result->st_ctime, &result->st_ctime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&info->LastWriteTime, &result->st_mtime, &result->st_mtime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&info->LastAccessTime, &result->st_atime, &result->st_atime_nsec);
+    result->st_nlink = info->NumberOfLinks;
+    result->st_ino = info->FileId.QuadPart;
+    /* bpo-37834: Only actual symlinks set the S_IFLNK flag. But lstat() will
+       open other name surrogate reparse points without traversing them. To
+       detect/handle these, check st_file_attributes and st_reparse_tag. */
+    result->st_reparse_tag = info->ReparseTag;
+    if (info->FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+        info->ReparseTag == IO_REPARSE_TAG_SYMLINK) {
+        /* set the bits that make this a symlink */
+        result->st_mode = (result->st_mode & ~S_IFMT) | S_IFLNK;
+    }
+    result->st_file_attributes = info->FileAttributes;
+    switch (info->DeviceType) {
+    case FILE_DEVICE_DISK:
+    case FILE_DEVICE_DISK_FILE_SYSTEM:
+    case FILE_DEVICE_VIRTUAL_DISK:
+    case FILE_DEVICE_DFS:
+    case FILE_DEVICE_CD_ROM:
+    case FILE_DEVICE_CD_ROM_FILE_SYSTEM:
+    case FILE_DEVICE_CONTROLLER:
+    case FILE_DEVICE_DATALINK:
+        break;
+    case FILE_DEVICE_CONSOLE:
+    case FILE_DEVICE_NULL:
+    case FILE_DEVICE_KEYBOARD:
+    case FILE_DEVICE_MODEM:
+    case FILE_DEVICE_MOUSE:
+    case FILE_DEVICE_PARALLEL_PORT:
+    case FILE_DEVICE_PRINTER:
+    case FILE_DEVICE_SCREEN:
+    case FILE_DEVICE_SERIAL_PORT:
+    case FILE_DEVICE_SOUND:
+        /* \\.\nul */
+        result->st_mode = (result->st_mode & ~S_IFMT) | _S_IFCHR;
+        break;
+    case FILE_DEVICE_NAMED_PIPE:
```

Comment:
From my earlier comment, we shouldn't arrive in `_Py_stat_basic_info_to_stat()` for device types that don't implement a filesystem or aren't mounted by a filesystem. That's because, in my experience, `FileBasicInformation` (file attributes and timestamps) is only implemented for files and directories in a filesystem. 

A filesystem type that mounts a volume will return `FILE_DEVICE_DISK` (e.g. sample[FAT32 source code](https://github.com/microsoft/Windows-driver-samples/blob/9e1a643093cac60cd333b6d69abc1e4118a12d63/filesys/fastfat/volinfo.c#L737-L795)) or `FILE_DEVICE_CD_ROM` (e.g. sample [CDFS source code](https://github.com/microsoft/Windows-driver-samples/blob/9e1a643093cac60cd333b6d69abc1e4118a12d63/filesys/cdfs/volinfo.c#L374-L427)). `FILE_DEVICE_DFS` is the distributed filesystem that combines multiple shares. A RAM disk device can use `FILE_DEVICE_VIRTUAL_DISK`, like the old RAM disk sample driver did, and the driver could implement a custom filesystem, but nowadays generally `FILE_DEVICE_DISK` or `FILE_DEVICE_CD_ROM` is used for a RAM disk (the ImDisk driver implements both).

`FILE_DEVICE_DISK_FILE_SYSTEM`, `FILE_DEVICE_CD_ROM_FILE_SYSTEM`, and `FILE_DEVICE_NETWORK_FILE_SYSTEM` are the device types for the base filesystem device, e.g. "\\\\?\\GlobalRoot\\FAT", "\\\\?\\GlobalRoot\UdfsCdRom", or "\\\\?\\UNC". `os.stat()` reports the first two as block devices (`S_IFBLK`), but `FILE_DEVICE_NETWORK_FILE_SYSTEM` isn't handled at all because `GetFileType()` returns `FILE_TYPE_UNKNOWN`. I don't think these low-level devices fit into the POSIX regime of `S_IFCHR` vs `S_IFBLK`, but they're more like `S_IFCHR` since reads and writes don't use a fixed block size. It's nearly a moot point since there's no useful reason to care. There isn't much to be done with such devices in user-mode applications. 





In principle, the stat info classes could be implemented for the named-pipe filesystem on the `FILE_DEVICE_NAMED_PIPE` device type, but in my tests NPFS doesn't implement `FileStatInformation`. I asked whether it's planned to implement stat info classes in NPFS.

Reply:
> From my earlier comment, we shouldn't arrive in _Py_stat_basic_info_to_stat() for device types that don't implement a filesystem or aren't mounted by a filesystem.

Ah, that's a very good point (which I missed). There were some tests that stopped failing with my change here though, so clearly something ends up in this path.

My switch basically mirrors `GetFileType`, so it should be the same as our existing code.

Other than leading people the wrong way when they read our code, is there anything _wrong_ with checking all the values? There's already a ton of stuff that doesn't map well into POSIX (all the error codes, for example), and I'd prefer our approximations at least be consistent with themselves, even if they aren't meaningful.

Reply:
> Other than leading people the wrong way when they read our code, is there anything wrong with checking all the values?

For the stat info classes, they could decide to specially handle some devices in the I/O manager, or implement a fallback in `GetFileInformationByName()`. I wouldn't want a device to be reported as a regular file instead of either `S_IFCHR` or `S_IFBLK`. It shouldn't be a problem as long as they only support the stat info classes for filesystem files and directories -- not for volume devices, filesystem devices, or controllers. (I don't think `FILE_DEVICE_DATALINK` is used by anything.)

Reply:
If they do end up supporting all devices for the stat info classes, it should be okay as long as they require that `FileAttributes` is set to `FILE_ATTRIBUTE_NORMAL` for a file that has no other attributes set. That's already required for `FileBasicInformation`, according to [\[MS-FSA\]](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-fsa/e4930504-ec22-4f48-b7c2-167e0f55d2f9). Devices would be distinguished by having `FileAttributes` set to 0.